File: original/Code/Combat/soldierfactorygameobj.cpp
#include "soldierfactorygameobj.h"
#include "basecontroller.h"
#include "simpledefinitionfactory.h"
#include "persistfactory.h"
#include "definitionmgr.h"
#include "combatchunkid.h"
#include "debug.h"
#include "basecontroller.h"
#include "combat.h"

////////////////////////////////////////////////////////////////
//	Hacks
////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////
//	Editable and persist factories
////////////////////////////////////////////////////////////////
SimplePersistFactoryClass	<SoldierFactoryGameObjDef,	CHUNKID_GAME_OBJECT_DEF_SOLDIER_FACTORY>							_SoldierFactoryGameObjDefPersistFactory;
SimplePersistFactoryClass	<SoldierFactoryGameObj,		CHUNKID_GAME_OBJECT_SOLDIER_FACTORY>								_SoldierFactoryGameObjPersistFactory;
DECLARE_DEFINITION_FACTORY (SoldierFactoryGameObjDef,	CLASSID_GAME_OBJECT_DEF_SOLDIER_FACTORY, "Soldier Factory")	_SoldierFactoryGameObjDefDefFactory;

////////////////////////////////////////////////////////////////
//	Save/Load constants
////////////////////////////////////////////////////////////////
enum
{
	CHUNKID_DEF_PARENT						=	0x02211153,
	CHUNKID_DEF_VARIABLES,

	MICROCHUNKID_DEF_UNUSED					= 1,
};

enum
{
	CHUNKID_PARENT								=	0x02211154,
	CHUNKID_VARIABLES,

	MICROCHUNKID_UNUSED						= 1,
};

////////////////////////////////////////////////////////////////
//
//	SoldierFactoryGameObjDef
//
////////////////////////////////////////////////////////////////
SoldierFactoryGameObjDef::SoldierFactoryGameObjDef (void)
{
	return ;
}

////////////////////////////////////////////////////////////////
//
//	~SoldierFactoryGameObjDef
//
////////////////////////////////////////////////////////////////
SoldierFactoryGameObjDef::~SoldierFactoryGameObjDef (void)
{
	return ;
}

////////////////////////////////////////////////////////////////
//
//	Get_Class_ID
//
////////////////////////////////////////////////////////////////
uint32
SoldierFactoryGameObjDef::Get_Class_ID (void) const	
{ 
	return CLASSID_GAME_OBJECT_DEF_SOLDIER_FACTORY;
}

////////////////////////////////////////////////////////////////
//
//	Create
//
////////////////////////////////////////////////////////////////
PersistClass *
SoldierFactoryGameObjDef::Create (void) const 
{
	SoldierFactoryGameObj *building = new SoldierFactoryGameObj;
	building->Init (*this);

	return building;
}

////////////////////////////////////////////////////////////////
//
//	Create
//
////////////////////////////////////////////////////////////////
bool
SoldierFactoryGameObjDef::Save (ChunkSaveClass &csave)
{
	csave.Begin_Chunk (CHUNKID_DEF_PARENT);		
		BuildingGameObjDef::Save (csave);
	csave.End_Chunk ();

	csave.Begin_Chunk (CHUNKID_DEF_VARIABLES);
	csave.End_Chunk ();
	return true;
}

////////////////////////////////////////////////////////////////
//
//	Load
//
////////////////////////////////////////////////////////////////
bool
SoldierFactoryGameObjDef::Load (ChunkLoadClass &cload)
{
	while (cload.Open_Chunk ())
	{
		switch (cload.Cur_Chunk_ID ())
		{
			case CHUNKID_DEF_PARENT:
				BuildingGameObjDef::Load (cload);
				break;

			case CHUNKID_DEF_VARIABLES:
				Load_Variables (cload);
				break;
	  
			default:
				Debug_Say (("Unrecognized Com Center Def chunkID\n"));
				break;

		}
		cload.Close_Chunk ();
	}

	return true;
}

////////////////////////////////////////////////////////////////
//
//	Load_Variables
//
////////////////////////////////////////////////////////////////
void
SoldierFactoryGameObjDef::Load_Variables (ChunkLoadClass &cload)
{
	while (cload.Open_Micro_Chunk ()) {
		
		/*switch (cload.Cur_Micro_Chunk_ID ())
		{
			default:
				Debug_Say (("Unrecognized Com Center Def Variable chunkID\n"));
				break;
		}*/

		cload.Close_Micro_Chunk();
	}

	return ;
}

////////////////////////////////////////////////////////////////
//
//	Get_Factory
//
////////////////////////////////////////////////////////////////
const PersistFactoryClass &
SoldierFactoryGameObjDef::Get_Factory (void) const 
{ 
	return _SoldierFactoryGameObjDefPersistFactory; 
}

////////////////////////////////////////////////////////////////
//
//	SoldierFactoryGameObj
//
////////////////////////////////////////////////////////////////
SoldierFactoryGameObj::SoldierFactoryGameObj (void)
{
	return ;
}

////////////////////////////////////////////////////////////////
//
//	~SoldierFactoryGameObj
//
////////////////////////////////////////////////////////////////
SoldierFactoryGameObj::~SoldierFactoryGameObj (void)
{
	return ;
}

////////////////////////////////////////////////////////////////
//
//	Get_Factory
//
////////////////////////////////////////////////////////////////
const PersistFactoryClass &
SoldierFactoryGameObj::Get_Factory (void) const 
{
	return _SoldierFactoryGameObjPersistFactory;
}

////////////////////////////////////////////////////////////////
//
//	Init
//
////////////////////////////////////////////////////////////////
void SoldierFactoryGameObj::Init (void)
{
	Init (Get_Definition ());
	return ;
}

////////////////////////////////////////////////////////////////
//
//	Init
//
////////////////////////////////////////////////////////////////
void
SoldierFactoryGameObj::Init (const SoldierFactoryGameObjDef &definition)
{
	BuildingGameObj::Init (definition);	
	return ;
}

////////////////////////////////////////////////////////////////
//
//	Get_Definition
//
////////////////////////////////////////////////////////////////
const SoldierFactoryGameObjDef &
SoldierFactoryGameObj::Get_Definition (void) const
{
	return (const SoldierFactoryGameObjDef &)BaseGameObj::Get_Definition ();
}

////////////////////////////////////////////////////////////////
//
//	Save
//
////////////////////////////////////////////////////////////////
bool
SoldierFactoryGameObj::Save (ChunkSaveClass &csave)
{
	csave.Begin_Chunk (CHUNKID_PARENT);
		BuildingGameObj::Save (csave);
	csave.End_Chunk ();

	csave.Begin_Chunk (CHUNKID_VARIABLES);
	csave.End_Chunk ();
	return true;
}

////////////////////////////////////////////////////////////////
//
//	Load
//
////////////////////////////////////////////////////////////////
bool
SoldierFactoryGameObj::Load (ChunkLoadClass &cload)
{
	while (cload.Open_Chunk ()) {
		switch (cload.Cur_Chunk_ID ()) {

			case CHUNKID_PARENT:
				BuildingGameObj::Load (cload);
				break;
								
			case CHUNKID_VARIABLES:
				Load_Variables (cload);
				break;

			default:
				Debug_Say (("Unrecognized Com Center chunkID\n"));
				break;
		}

		cload.Close_Chunk();
	}

	return true;
}

////////////////////////////////////////////////////////////////
//
//	Load_Variables
//
////////////////////////////////////////////////////////////////
void
SoldierFactoryGameObj::Load_Variables (ChunkLoadClass &cload)
{
	while (cload.Open_Micro_Chunk ()) {
		
		/*switch (cload.Cur_Micro_Chunk_ID ())
		{
			default:
				Debug_Say (("Unrecognized Com Center Variable chunkID\n"));
				break;
		}*/

		cload.Close_Micro_Chunk();
	}

	return ;
}

////////////////////////////////////////////////////////////////
//
//	On_Destroyed
//
////////////////////////////////////////////////////////////////
void
SoldierFactoryGameObj::On_Destroyed (void)
{
	BuildingGameObj::On_Destroyed ();

	//
	//	Switch off the radar for each player on this team
	//
	if (BaseController != NULL && CombatManager::I_Am_Server ()) {
		BaseController->Set_Can_Generate_Soldiers (false);
	}

	return ;
}

////////////////////////////////////////////////////////////////
//
//	CnC_Initialize
//
////////////////////////////////////////////////////////////////
void
SoldierFactoryGameObj::CnC_Initialize (BaseControllerClass *base)
{
	BuildingGameObj::CnC_Initialize (base);

	//
	//	Let the base know it can generate soldiers
	//
	if (BaseController != NULL) {
		BaseController->Set_Can_Generate_Soldiers (true);
	}

	return ;
}

File: kotlin-server/server/src/main/kotlin/ccr/server/net/SoldierFactoryGameObj.kt
package ccr.server.net

import ccr.server.defs.SoldierFactoryGameObjDef

// C++: SoldierFactoryGameObj : public BuildingGameObj (soldierfactorygameobj.h)
// Hierarchy: NetworkObject → BaseGameObj → ScriptableGameObj → DamageableGameObj →
//            BuildingGameObj → SoldierFactoryGameObj
//
// SoldierFactoryGameObj adds no new member fields. Its only behaviour on top of
// BuildingGameObj is to register/revoke the "can generate soldiers" capability on
// the base controller via CnC_Initialize and On_Destroyed.
class SoldierFactoryGameObj() : BuildingGameObj() {

    // Secondary constructor for tests — bypasses Init() / definition pipeline.
    constructor(
        definitionId: Int,
        position: ccr.math.Vector3 = ccr.math.Vector3(),
        sphereCenter: ccr.math.Vector3 = ccr.math.Vector3(),
        sphereRadius: Float = 50f,
        health: Float = 0f,
        isDestroyed: Boolean = false,
        isPowerOn: Boolean = true,
    ) : this() {
        definition = ccr.server.defs.BaseGameObjDef(
            name = "soldierfactory_$definitionId", id = definitionId.toUInt(), chunkId = 0u
        )
        this.position = position
        this.collectionSphere = ccr.server.level.Sphere(sphereCenter, sphereRadius)
        this.isDestroyed = isDestroyed
        this.isPowerOn = isPowerOn
        defenseObject.health = health
    }

    // C++: virtual void Init(void) — calls Init(Get_Definition())
    override fun init() {
        init(getDefinition())
    }

    // C++: void Init(const SoldierFactoryGameObjDef& definition)
    fun init(definition: SoldierFactoryGameObjDef) {
        // C++: BuildingGameObj::Init(definition)
        super.init(definition)
    }

    // C++: const SoldierFactoryGameObjDef& Get_Definition() const
    override fun getDefinition(): SoldierFactoryGameObjDef =
        definition as SoldierFactoryGameObjDef

    // C++: SoldierFactoryGameObj* As_SoldierFactoryGameObj() { return this; }
    // → handled by Kotlin type system

    // -------------------------------------------------------------------------
    // C++: virtual bool Save(ChunkSaveClass& csave)
    // -------------------------------------------------------------------------
    override fun save(csave: ChunkSaveClass): Boolean {
        csave.beginChunk(CHUNKID_PARENT)
        super.save(csave)
        csave.endChunk()

        csave.beginChunk(CHUNKID_VARIABLES)
        csave.endChunk()

        return true
    }

    // -------------------------------------------------------------------------
    // C++: virtual bool Load(ChunkLoadClass& cload)
    // -------------------------------------------------------------------------
    override fun load(cload: ChunkLoadClass): Boolean {
        while (cload.openChunk()) {
            when (cload.curChunkId) {
                CHUNKID_PARENT    -> super.load(cload)
                CHUNKID_VARIABLES -> loadVariables(cload)
                else -> error("Unrecognized SoldierFactoryGameObj chunk ID: ${cload.curChunkId}")
            }
            cload.closeChunk()
        }
        return true
    }

    // C++: void Load_Variables(ChunkLoadClass& cload)
    // The switch block is entirely commented out in C++ — just loop and close.
    private fun loadVariables(cload: ChunkLoadClass) {
        while (cload.openMicroChunk()) {
            // C++: switch is commented out — no micro-chunks are defined
            cload.closeMicroChunk()
        }
    }

    // -------------------------------------------------------------------------
    // C++: void CnC_Initialize(BaseControllerClass* base)
    // -------------------------------------------------------------------------
    override fun cncInitialize(base: BaseControllerClass) {
        // C++: BuildingGameObj::CnC_Initialize(base)
        super.cncInitialize(base)

        // C++: if (BaseController != NULL) BaseController->Set_Can_Generate_Soldiers(true)
        if (baseController != null) {
            baseController!!.setCanGenerateSoldiers(true)
        }
    }

    // -------------------------------------------------------------------------
    // C++: virtual void On_Destroyed()
    // -------------------------------------------------------------------------
    override fun onDestroyed() {
        // C++: BuildingGameObj::On_Destroyed()
        super.onDestroyed()

        // C++: if (BaseController != NULL && CombatManager::I_Am_Server())
        //          BaseController->Set_Can_Generate_Soldiers(false)
        // (always server — omit I_Am_Server guard per guide)
        if (baseController != null) {
            baseController!!.setCanGenerateSoldiers(false)
        }
    }

    companion object {
        // C++: enum { CHUNKID_PARENT = 0x02211154, CHUNKID_VARIABLES, ... }
        private const val CHUNKID_PARENT    = 0x02211154
        private const val CHUNKID_VARIABLES = 0x02211155

        // C++: MICROCHUNKID_UNUSED = 1 (switch entirely commented out — unused)
        @Suppress("unused")
        private const val MICROCHUNKID_UNUSED = 1
    }
}
